src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -203,15 +203,13 @@ def remap(
 
     __call__ = remap  # type: ignore[assignment]
 
-    def restrict(self, index: common.AnyIndexSpec) -> common.Field | core_defs.ScalarT:
+    def restrict(self, index: common.AnyIndexSpec) -> common.Field:
         new_domain, buffer_slice = self._slice(index)
-
         new_buffer = self.ndarray[buffer_slice]
-        if len(new_domain) == 0:
-            # TODO: assert core_defs.is_scalar_type(new_buffer), new_buffer
-            return new_buffer  # type: ignore[return-value] # I don't think we can express that we return `ScalarT` here
-        else:
-            return self.__class__.from_array(new_buffer, domain=new_domain)
+        if new_domain.ndim == 0:
+            assert self.array_ns.dtype(new_buffer) in core_defs.SCALAR_TYPES
+            return self._scalar_to_field(new_buffer)  # type: ignore[return-value, arg-type]
```

Comment:
This is returning an array not a field. I think `NdArrayField.from_array()` could be used here, since it should work for both scalars and ndarrays.